### PR TITLE
CB-9365 Add support for 'vibrateWithPattern' to Windows Phone 8.1 / W…

### PR DESCRIPTION
**Pull request by daserge**
…indows 10

[Jira issue](https://issues.apache.org/jira/browse/CB-9365)

### CODE REVIEW

File: src/windows/VibrationProxy.js [hunk 1]
```diff
@@ -60,12 +61,43 @@ if (VibrationDevice) {
...
             }
         }, 
-        vibrateWithPattern: function(success, fail, args) {
-            // TODO: Implement with setTimeout.
-            fail('"vibrateWithPattern" is unsupported by this platform.');
+        vibrateWithPattern: function (success, fail, args) {
+            // Cancel current vibrations first
+            module.exports.cancelVibration(function () {
+                // Removes the first zero
+                if (args[0] && args[0][0] === 0) {
```
Comment (nikhilkh) [inline src/windows/VibrationProxy.js]: Can we factor out args[0] to a more meaningful variable name?

File: README.md [hunk 2]
```diff
@@ -113,7 +113,9 @@ Vibrates the device with a given pattern
 
 ####Windows Quirks
 
-- vibrate(pattern) falls back on vibrate with default duration
+- `repeat` parameter is not supported
```
Comment (nikhilkh) [inline README.md]: Why is repeat not supported?

File: src/windows/VibrationProxy.js [hunk 3]
```diff
@@ -60,12 +61,43 @@ if (VibrationDevice) {
...
+                    args[0].shift(0);
+                }
+
+                patternChainPromise = args[0].reduce(function (previousValue, currentValue, index) {
```
Comment (nikhilkh) [inline src/windows/VibrationProxy.js]: Do we need to test for presence of args[0]

**Comment by nikhilkh** [comment]
Should any mobilespec tests be updated now that this is supported for Windows? Or is mobilespec not testing this because its not supported on emulators?

File: src/windows/VibrationProxy.js [hunk 4]
```diff
@@ -60,12 +61,43 @@ if (VibrationDevice) {
...
+                                }, [currentValue]);
+
+                                setTimeout(function () {
+                                    complete();
+                                }, parseInt(currentValue, 10));
+                            });
+                        });
+                    } else {
+                        return previousValue.then(function () {
+                            return new WinJS.Promise(function (complete) {
```
Comment (robpaveza) [inline src/windows/VibrationProxy.js]: Prefer `WinJS.Promise.timeout(parseInt(currentValue, 10))` to `new WinJS.Promise( ... setTimeout(...))`.

File: src/windows/VibrationProxy.js [hunk 5]
```diff
@@ -60,12 +61,43 @@ if (VibrationDevice) {
...
+                                }, parseInt(currentValue, 10));
+                            });
+                        });
+                    }
+                }, WinJS.Promise.as());
+            }, fail);
         },
         cancelVibration: function(success, fail, args) {
             try {
+                patternChainPromise && patternChainPromise.cancel();
```
Comment (robpaveza) [inline src/windows/VibrationProxy.js]: If you're currently vibrating, will the vibration itself also be cancelled?

File: src/windows/VibrationProxy.js [hunk 6]
```diff
@@ -60,12 +61,43 @@ if (VibrationDevice) {
...
+                }
+
+                patternChainPromise = args[0].reduce(function (previousValue, currentValue, index) {
+                    if (index % 2 === 0) {
+                        return previousValue.then(function () {
+                            return new WinJS.Promise(function (complete, error) {
+                                module.exports.vibrate(function () {}, function(err) {
+                                    console.error(err);
+                                    error(err);
+                                }, [currentValue]);
```
Comment (robpaveza) [inline src/windows/VibrationProxy.js]: We make a lot of references to `currentValue`.  Can we just parse it once and save the integer form?  Maybe we can opportunistically parse all of them on input and if any parse return NaN then we fail fast.

File: src/windows/VibrationProxy.js (hunk 5 above)
Comment (daserge) [inline src/windows/VibrationProxy.js]: Yes, it will be cancelled by the next line:
```javascript
VibrationDevice.getDefault().cancel();
```

**Comment by daserge** [comment]
@nikhilkh, @robpaveza thanks for review!
Please take a look once more - I have addressed your comments.

**Pull request merged**